javatests/arcs/android/systemhealth/testapp/StorageCore.kt

```diff
@@ -396,51 +429,80 @@ class StorageCore(val context: Context, val lifecycle: Lifecycle) {
-                if (exception !is InterruptedException) {
-                    val timestamp = System.currentTimeMillis()
-                    val taskType = controllers.getOrNull(taskId)?.taskType?.name ?: ""
-                    val msg = "$taskType #$taskId: $exception"
-                    log.error { msg }
-                    exception.printStackTrace()
-                    taskManagerEvents.writer.withLock {
-                        taskManagerEvents.queue.add(
-                            TaskEvent(TaskEventId.EXCEPTION, timestamp, desc = msg))
-                    }
+    private fun performanceExceptionHandler(taskId: Int) =
+        CoroutineExceptionHandler { _, exception ->
+            // This is the innermost thread-wise exception handler monitoring all exceptions
+            // that happened in coroutine context against the outermost one which is installed
+            // at ScheduledThreadPoolExecutor.
+            if (exception !is InterruptedException) {
+                val timestamp = System.currentTimeMillis()
+                val taskType = controllers.getOrNull(taskId)?.taskType?.name ?: ""
+                val msg = "$taskType #$taskId: $exception"
+                log.error { msg }
+                exception.printStackTrace()
+                taskManagerEvents.writer.withLock {
+                    taskManagerEvents.queue.add(
+                        TaskEvent(TaskEventId.EXCEPTION, timestamp, desc = msg))
+                }
+
+                // Cancel all pending tasks and forbid new tasks.
+                controllers.getOrNull(taskId)?.future?.cancel(false)
+            }
 
-                    // Cancel all pending tasks and forbid new tasks.
-                    controllers.getOrNull(taskId)?.future?.cancel(false)
+            // Don't ignore the exception being received.
+            // The single task thread will be terminated upon the unhandled exception.
+            throw exception
+        }
+
+    private fun stabilityExceptionHandler(taskId: Int) =
+        CoroutineExceptionHandler { _, exception ->
+
+            if (exception !is StorageClientException) {
+                val timestamp = System.currentTimeMillis()
+                val taskType = controllers.getOrNull(taskId)?.taskType?.name ?: ""
+                val msg = "$taskType #$taskId: $exception"
+                taskManagerEvents.writer.withLock {
+                    taskManagerEvents.queue.add(
+                        TaskEvent(TaskEventId.EXCEPTION, timestamp, desc = msg))
                 }
 
-                // Don't ignore the exception being received.
-                // The single task thread will be terminated upon the unhandled exception.
-                throw exception
+                // Cancel all pending tasks and forbid new tasks.
+                controllers.getOrNull(taskId)?.future?.cancel(false)
+
+                earlyExit = true
+                // Ignore the exception as it's reported.
             }
+        }
```

Comment:
Maybe throw an InterruptedException here to kill the task thread at once due to the injected crash or the unhandled exception, otherwise the client can still continue to execute after the unhandled exception or crash which should be avoided otherwise we may see a bunch of exception bombs in a short time.

Reply:
Done.